addons/l10n_in/models/account_invoice.py

```diff
@@ -71,23 +71,49 @@ def _compute_l10n_in_state_id(self):
-        for move in in_moves:
-            state_id = move.l10n_in_state_id.id
-            country_id = move.l10n_in_state_id.country_id.id
-            if move.l10n_in_state_id.l10n_in_tin == '96':  # Foreign Country
-                other_country = other_country or self.env['res.country'].search([('code', '!=', 'IN')], limit=1)
-                country_id = other_country.id
-                state_id = False
-            virtual_partner = self.env['res.partner'].new({
-                'state_id': state_id,
-                'country_id': country_id,
-            })
-            move.fiscal_position_id = self.env['account.fiscal.position'].with_company(
-                move.company_id
-            )._get_fiscal_position(virtual_partner)
-        super(AccountMove, self - in_moves)._compute_fiscal_position_id()
+        computed_ids = []
+        for move in self:
+            if (
+                move.country_code != 'IN'
+                or not move.is_invoice(include_receipts=True)
+                or move.partner_shipping_id.property_account_position_id
+                or move.partner_id.property_account_position_id
+            ):
+                continue
+            dummy_state_id = (
+                move.l10n_in_gst_treatment == 'special_economic_zone'
+                and self.env.ref('l10n_in.state_in_oc')
+                or False
+            )
+            # If move type is sale, compare place of supply with company state
+            if not dummy_state_id and move.is_sale_document(include_receipts=True):
+                dummy_state_id = move.l10n_in_state_id
+            # If move type is purchase, compare place of supply with vendor state
+            elif (
+                not dummy_state_id
+                and move.is_purchase_document(include_receipts=True)
+                and move.partner_id.country_id.code == 'IN'
+            ):
+                pos_state_id = move.l10n_in_state_id
+                # Vendor's place of supply is within his home state and not same as the company state
+                if pos_state_id == move.partner_id.state_id != move.company_id.state_id:
+                    dummy_state_id = move.company_id.state_id
+                # Vendor and Company are in same state but place of supply is other state
+                elif (
+                    pos_state_id != move.partner_id.state_id == move.company_id.state_id
+                    or pos_state_id.l10n_in_tin == '96'
+                ):
+                    dummy_state_id = pos_state_id
+            if dummy_state_id:
+                virtual_partner = self.env['res.partner'].new({
+                    'state_id': dummy_state_id.id,
+                    'country_id': dummy_state_id.country_id.id,
+                })
+                move.fiscal_position_id = self.env['account.fiscal.position'].with_company(
+                    move.company_id
+                )._get_fiscal_position(virtual_partner)
+                computed_ids.append(move.id)
```

Comment:
Why not just use a recordset?
in_moves |= move or something

Reply:
We should avoid using operation on recordset in a loop. We have done operator overloading in the base but the recordsets are frozen datatype. Whenever performing any operation on recordsets it will return a new recordset by browsing the `ids` with complexity of `O(N)` and using it in a loop makes it `O(N^2)`

https://github.com/odoo/odoo/blob/dee2a222b72c86e20e68489c9cf38d8683dee832/odoo/orm/models.py#L6772-L6783